sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,802 @@
+
+template <typename dataT,
+          int dimensions,
+          access::mode accessmode,
+          access::target accessTarget,
+          access::placeholder isPlaceholder,
+          typename PropertyListT = ext::oneapi::property_list<>>
+class accessor {/* ... */};
+
+} // namespace sycl
+```
+
+Typical usage would look like this (showing a hypothetical property named
+`foo`):
+
+```
+using sycl;
+using sycl::ext::oneapi;
+
+accessor acc(buf, cgh, property_list{no_alias_v, foo_v<32>});
+```
+
+As before, the header file represents the properties with an internal C++
+attribute, where the initial parameters are the names of the properties and
+the subsequent parameters are the property values.
+
+```
+namespace sycl {
+
+template <typename dataT,
+          int dimensions,
+          access::mode accessmode,
+          access::target accessTarget,
+          access::placeholder isPlaceholder,
+          typename PropertyListT = ext::oneapi::property_list<>>
+class accessor {/* ... */};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename dataT,
+          int dimensions,
+          access::mode accessmode,
+          access::target accessTarget,
+          access::placeholder isPlaceholder,
+          typename ...Props>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_kernel_parameter_attributes(
+    Props::meta_name..., Props::meta_value...
+    )]]
+#endif
+  accessor<dataT,
+           dimensions,
+           accessmode,
+           accessTarget,
+           isPlaceholder,
+           property_list<Props...>> {/*...*/};
+
+} // namespace sycl
+```
+
+Illustrating this with the previous example:
+
+```
+namespace sycl {
+
+template </* ... */> class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_kernel_parameter_attributes(
+    "sycl-no-alias",  // Name of first property
+    "sycl-foo",       // Name of second property
+    nullptr,          // First property has no parameter
+    32                // Value of second property
+    )]]
+#endif
+  accessor</* ... */> {/* ... */};
+
+} // namespace sycl
+```
+
+As the name of the C++ attribute suggests, the device compiler front-end uses
+the attribute only when the decorated type is the type of a kernel argument,
+and it silently ignores the attribute when the class is used in any other way.
+
+When the device compiler front-end creates a kernel argument in this way, it
+adds one LLVM IR attribute to the kernel function's parameter for each property
+in the list.  For example, this can be done by calling
+[`Function::addParamAttrs(unsigned ArgNo, const AttrBuilder &)`][7].  As
+before, the IR attributes are added as strings, so the front-end must convert
+the property value to a string if it is not already a string.
+
+[7]: <https://llvm.org/doxygen/classllvm_1_1Function.html#a092beb46ecce99e6b39628ee92ccd95a>
+
+**TODO**: What happens when a "sycl special class" object is captured as a
```

Comment:
Thanks, this is good to know.  Since this design attempts to provide a general framework for properties, though, I think there is still an issue in the general case.

Reply:
This TODO is removed as of d0622a60026db7e2ff2d9a69ac8cdfb1ad05e917.  The general solution is similar to what we do with BufferLocationProperty and NoAliasProperty.